### PR TITLE
update engine-jupyter chart

### PR DESCRIPTION
**Pull request by smola**
* renamed from spark-api to engine-jupyter.
  
* use jupyter-spark Docker image and use --packages
   to download engine, this seems to work with cluster.    
   mode whereas engine-jupyter Docker image does not.
    
* make bblfshd optional.
  
* updated bblfshd driver installation mechanism to be
  the same as in bblfshd chart.

Signed-off-by: Santiago M. Mola <santi@mola.io>

### CODE REVIEW

File: engine-jupyter/templates/deployment.yaml [hunk 1]
```diff
@@ -37,22 +38,25 @@ spec:
...
+{{- if .Values.bblfshd.enabled }}
+        - name: bblfshd
+          image: "{{ .Values.bblfshd.image.repository }}:{{ .Values.bblfshd.image.tag }}"
+          imagePullPolicy: {{ .Values.bblfshd.image.pullPolicy }}
           securityContext:
             privileged: true
+          command: ["/bin/sh"]
+          # Yes, at this point we probably want to add some command line argument to
+          # bblfshd to install a set of drivers at start up.
+          args: ["-c", "bblfshd & pid=$! ; sleep 1 && bblfshctl driver install --recommended && kill $pid && exec bblfshd"]
```
[the remark this answers is not in the document]
Comment (rporres) [reply engine-jupyter/templates/deployment.yaml]: This also has another problem. If you have multiple pods and one gets restarted you could end up with pods with different driver versions

File: engine-jupyter/values.yaml [hunk 2]
```diff
@@ -42,12 +47,11 @@ jupyter:
...
+  enabled: true
   image:
     repository: bblfsh/bblfshd
-    tag: v2.0.0
+    tag: v2.4.2
     pullPolicy: IfNotPresent
-  driversPath: /var/lib/bblfshd
-  #driversDisk: bblfsh-drivers
-  resources: {}
+  resources: {}
```
Comment (rporres) [inline engine-jupyter/values.yaml]: nit: missing newline at end of file

**Review by rporres — COMMENTED** [review]
This chart was written to be deployed in GCE when we were trying a first approach for the engine playground, that's why it had certain volume assumptions. I finally had to use a different approach. Great that it's been useful anyway.

A few comments:

* bblfsh driver installation is tricky if you want to do it correctly. I've posted some suggestions, but we can leave them for a further version if it is out of the scope (as long as we have some TODO in the code)

* If you want to store your notebooks you can use a hostPath (with a mandatory nodeSelector) or a PersistentVolumeClaim. Unfortunately local PVC are only beta in k8s 1.10 so the most general way to code this would be giving both chances

File: engine-jupyter/templates/deployment.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (smola) [reply engine-jupyter/templates/deployment.yaml]: Changed. Copied what bblfshd chart has.

File: engine-jupyter/templates/deployment.yaml [hunk 3]
```diff
@@ -15,18 +15,31 @@ spec:
         app: {{ template "name" . }}
         release: {{ .Release.Name }}
     spec:
+      volumes:
+        - name: notebooks-dir
```
Comment (rporres) [inline engine-jupyter/templates/deployment.yaml]: I'd give this as an option as it limits how this is deployed. And it closes the door to use this with persistent volume claims which is the standard way of describing volumes in k8s. I'd write something like:

```
# TODO: Add PVC support
{{- if .Values.notebooks.volumes }}
  volumes:
    - name: notebooks-dir
  {{- if .Values.notebooks.volumes.hostPath }}
      hostPath:
        path: {{ .Values.notebooks.volumes.hostPath }}
  {{- end }}
{{- end }}

File: engine-jupyter/templates/deployment.yaml [hunk 4]
```diff
@@ -35,24 +48,34 @@ spec:
             httpGet:
               path: /
               port: {{ .Values.service.internalPort }}
+          volumeMounts:
```
Comment (rporres) [inline engine-jupyter/templates/deployment.yaml]: As comented above, I'd give the opportunity of not saving state if no volume is given and add the same condition:

```
{{- if .Values.notebooks.volumes }}
```

File: engine-jupyter/values.yaml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    # kubernetes.io/tls-acme: "true"
+  tls:
+    # Secrets must be manually created in the namespace.
+    # - secretName: chart-example-tls
+    #   hosts:
+    #     - chart-example.local
+notebooks:
+  # Required home path for jupyter to store notebooks.
+  # hostPath: /data-a/notebooks
+  hostPath:
```
Comment (rporres) [inline engine-jupyter/values.yaml]: To be consistent with the above suggestion, I'd organize this as follows

```
notebooks:
  volumes: {}
    # hostPath: /data-a/notebooks
```

**Review by rporres — CHANGES_REQUESTED** [review]
I'd rather add persistence as an option. Suggested syntax (untested, though)

File: engine-jupyter/templates/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  replicas: {{ .Values.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "name" . }}
+        release: {{ .Release.Name }}
+    spec:
+      volumes:
+        # TODO: Add PVC support
+        {{ if .Values.notebooks.volumes }}
```
Comment (rporres) [inline engine-jupyter/templates/deployment.yaml]: nit: please use `{{- if` to avoid new lines

**Review by rporres — CHANGES_REQUESTED** [review]
There are a few `{{ if ...` and `{{ end  }}` that create unnecessary empty lines

**Comment by smola** [comment]
Closing, too outdated now.

**Pull request closed without merging**